pylint/checkers/unicode.py

```diff
@@ -0,0 +1,713 @@
+    else:
+        # UTF 16 and UTF-32 are sucking, so first get all the bytes in memory
+        content = b"".join(line for line in steam)
+
+        new_line = _cached_encode_search("\n", codec)
+
+        # Now we split the line by the real new line in the correct encoding
+        # we can't use split as it would strip the \n that we need
+        start = 0
+        while True:
+            pos = content.find(new_line, start)
+            if pos >= 0:
+                yield content[start : pos + len(new_line)]
+            else:
+                # Yield the rest and finish
+                yield content[start:]
+                break
+
+            start = pos + len(new_line)
+
+
+def extract_codec_from_bom(first_line: bytes) -> str:
+    """Try to extract the codec (unicode only) by checking for the BOM
+
+    For details about BOM see https://unicode.org/faq/utf_bom.html#BOM
+
+    Args:
+        first_line: the first line of a file
+
+    Returns:
+        a codec name
+
+    Raises:
+        ValueError: if no codec was found
+    """
+    for bom, codec in BOM_SORTED_TO_CODEC.items():
+        if first_line.startswith(bom):
+            return codec
+
+    raise ValueError("No BOM found. Could not detect Unicode codec.")
+
+
+class UnicodeChecker(BaseChecker):
+    """Check characters that could be used to hide bad code to humans
+
+    This includes:
+
+    - Bidirectional Unicode (see https://trojansource.codes/)
+
+    - Bad ASCII characters (see PEP672)
+
+        If a programmer requires to use such a character they should use the escaped
+        version, that is also much easier to read and does not depend on the editor used.
+
+    The Checker also includes a check that UTF-16 and UTF-32 are not used to encode
+    Python files.
+
+    At the time of writing Python supported only UTF-8. See
+    https://stackoverflow.com/questions/69897842/ and https://bugs.python.org/issue1503789
+    for background.
+    """
+
+    __implements__ = interfaces.IRawChecker
+    priority = -1
+
+    name = "unicode_checker"
+
+    msgs = {
+        "E2401": (
+            # This error will be only displayed to users once Python Supports
+            # UTF-16/UTF-32 (if at all)
+            "UTF-16 and UTF-32 aren't backward compatible. Use UTF-8 instead",
+            "invalid-unicode-codec",
+            (
+                "For compatibility use UTF-8 instead of UTF-16/UTF-32."
+                "See also https://bugs.python.org/issue1503789 for a history "
+                "of this issue. And "
+                "https://softwareengineering.stackexchange.com/questions/102205/should-utf-16-be-considered-harmful "
+                "for some possible problems when using UTF-16 for instance."
+            ),
+        ),
+        "E2402": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution.",
```

Comment:
```suggestion
            "special unicode characters could means that someone is trying to hide"
            " what the code is really doing from you",
            "bidirectional-unicode",
            "bidirectional unicode are typically not displayed and can affect the meaning"
            " of the source code. Something deceptive or nefarious might be going on."
            "Do you trust this code ?"
```

Reply:
Hmm. I think some more explanation is required there. 
On the one hand we would like to enable (as you wrote before) users of [RTL languages](https://en.wikipedia.org/wiki/Right-to-left_script) to write at least comments in their native language but at the same time make them aware that code could look different as they would expect them to be executed.

And for everybody that does not use a RTL language this should be warning that some unwanted character is within their text. 

I came up with the following:
```python
        "E2402": (
            (
                "special unicode characters only required to display right-to-left "
                "scripts (i.e. Chinese, Japanese, Arabic, Hebrew, ...). "
                "But they could lead to code that is executed differently as what you "
                "would expect looking at it in your editor."
            ),
            "bidirectional-unicode",
            (
                "bidirectional unicode are typically not displayed characters required "
                "to display right-to-left (RTL) script "
                "(i.e. Chinese, Japanese, Arabic, Hebrew, ...) correctly. "
                "If you wrote text (i.e. in a comment) in a RTL language make, ensure "
                "you don't mix code and your RTL text before disabling this warning!"
                "If you did not write text in a RTL language, remove the special "
                "characters, as they could be used to trick you into executing code,"
                "that does something else as it looks like.\n"
                "Do you trust this code ?\n"
                "More Information:\n"
                "https://en.wikipedia.org/wiki/Bidirectional_text\n",
                "https://trojansource.codes/"
            ),
        ),
```
Maybe @Pierre-Sassoulas comment here before I commit.

Reply:
The first one must be short and to the point as it's going to be printed in full every time. Let me know if this is better:
```suggestion
            "right-to-left unicode characters could means underhanded/nefarious code"
            " if you're not writing right to left yourself.",
            "bidirectional-unicode",
            (
                "bidirectional unicode are typically not displayed characters required "
                "to display right-to-left (RTL) script (i.e. Chinese, Japanese, Arabic, Hebrew, ...) correctly. "
                "So can you trust this code ? If you did not write it or your language is not RTL,"
                " remove the special characters, as they could be used to trick you into executing code,"
                "that does something else than what it looks like.\n"
                "More Information:\n"
                "https://en.wikipedia.org/wiki/Bidirectional_text\n",
                "https://trojansource.codes/"
            ),
```

Reply:
We actually checking not only for for right-to-left unicode characters but also left-to-right that could cause strange displays when left unmatched.

"if you not writing right to left" could be quite confusing if you don't know that this is referring to a script system.



For the last one: I would like to include a hint for native speakers that use RTL scripts: Because even if you use them correctly, as described in the example in PEP8, they can lead to issues.

Next suggestion:
```python
            "contains control characters that can lead to code executed differently as displayed",
            "bidirectional-unicode",
            (
                "bidirectional unicode are typically not displayed characters required "
                "to display right-to-left (RTL) script (i.e. Chinese, Japanese, Arabic, Hebrew, ...) correctly. "
                "So can you trust this code? Are you sure it displayed correctly in all editors? 
                " If you did not write it or your language is not RTL,"
                " remove the special characters, as they could be used to trick you into executing code,"
                "that does something else than what it looks like.\n"
                "More Information:\n"
                "https://en.wikipedia.org/wiki/Bidirectional_text\n",
                "https://trojansource.codes/"
            ),
```
1st following KISS as much as possible.
3rd trying to be verbose as possible.

Reply:
Ok, I understand better :) What about:
```diff
- "contains control characters that can lead to code executed differently as displayed",
+ "contains control characters that can permit underhanded constructs that will execute"
+ " differently as displayed",
```

Reply:
I am not a native English speaker but "underhanded constructs" sounds strange. But I know what you mean.
IMO we should not only warn about the possible security issue but also about issues in displaying in general.
The go hand in hand.
Maybe without the "underhanded constructs" ?
Or maybe @DanielNoord has a formulation idea for "underhanded constructs" ?

Reply:
``underhanded code``, ``obfuscated code``,  maybe ?

Reply:
I go for obfuscated.